### PR TITLE
Refactor tab to serialize to storage

### PR DESCRIPTION
**Pull request by jonathanKingston**
**Reviewer:** @shakyShane 

## Description:

1. Create a new class to manage the state of tabs and sites.
These need writing to session storage to allow us to restore their state after the service worker is reinitialized.

2. Uses getters and setters in to store into the new class from tab and site.

3. Changing the tab and site shape to have getters and setters will break the consumers of the current tab serialization also (we serialize over message passing to different contexts which turns classes into plain objects).
So additionally we will need to manually serialize into this plain object format for sending over to different contexts; this will allow autofill, content-scope-scripts and the dashboard to retain their current usage of tab and site data.

https://app.asana.com/0/0/1202914947727988/f

## Automated tests:
- [ ] Unit tests
- [ ] Integration tests

###### Reviewer Checklist:
- [ ] **Ensure the PR solves the problem**
- [ ] **Review every line of code**
- [ ] **Ensure the PR does no harm by testing the changes thoroughly**
- [ ] **Get help if you're uncomfortable with any of the above!**
- [ ] Determine if there are any quick wins that improve the implementation

###### PR Author Checklist:
- [ ] Get advice or leverage existing code
- [ ] Agree on technical approach with reviewer (if the changes are nuanced)
- [ ] Ensure that there is a testing strategy (and documented non-automated tests)
- [ ] Ensure there is a documented monitoring strategy (if necessary)
- [ ] Consider systems implications

### CODE REVIEW

File: shared/js/background/classes/tab-state.js [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+
+    /**
+     * Processes the storage queue in order.
+     */
+    async processQueue () {
+        if (!this.processing) {
+            while (this.taskQueue.length > 0) {
+                this.processing = true
+                const task = this.taskQueue.shift()
+                await task()
```
Comment (shakyShane) [inline shared/js/background/classes/tab-state.js:123]: should errors cause processing to halt for all?

This example wont run `console.log(3)`

```js

const Storage = new StorageInstance()
Storage.addTask(async () => {
    console.log('1')
})
Storage.addTask(async () => {
    console.log('2')
    throw new Error("oops!")
})
Storage.addTask(async () => {
    console.log('3')
    done()
})
```

**Review by shakyShane — CHANGES_REQUESTED** [review]
Just a few small comments :)

File: shared/js/background/classes/legacy-tab-transfer.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        }
+        if (isStructuredCloneable(value)) {
+            out[key] = structuredClone(value)
+        } else {
+            out[key] = cloneClassObject(value)
+        }
+    }
+    if (hasModifiedPrototype(object)) {
+        const objectDescriptors = Object.getOwnPropertyDescriptors(Object.getPrototypeOf(object))
+        // const getters = Object.entries(objectDescriptors).filter(([key, descriptor]) => typeof descriptor.get === 'function')
```
[the remark this answers is not in the document]
Comment (jonathanKingston) [reply shared/js/background/classes/legacy-tab-transfer.js:52]: Weird I thought I removed this :)

File: shared/js/background/classes/tab.es6.js [hunk 3]
```diff
@@ -29,37 +29,187 @@ const Site = require('./site.es6')
...
+
+    get trackers () {
+        return this._tabState.trackers
+    }
+
+    get url () {
+        return this._tabState.url
+    }
+
+    set url (url) {
```
[the remark this answers is not in the document]
Comment (jonathanKingston) [reply shared/js/background/classes/tab.es6.js:103]: I'm wondering if we should remove all the types here. TS actually knows them all safely.

[the remark this answers is not in the document]
Comment (jonathanKingston) [reply shared/js/background/classes/tab.es6.js:103]: So I trimmed all of them and made sure tab-state instead had the definitions. I proved locally that we can have nice types in these methods with tsc and also typedoc (which we intend to use probably)

File: shared/js/background/classes/tab.es6.js [hunk 4]
```diff
@@ -29,37 +29,187 @@ const Site = require('./site.es6')
...
+     */
+    set cleanAmpUrl (url) {
+        this._tabState.setValue('cleanAmpUrl', url)
+    }
+
+    get status () {
+        return this._tabState.status
+    }
+
+    set status (value) {
```
[the remark this answers is not in the document]
Comment (jonathanKingston) [reply shared/js/background/classes/tab.es6.js:203]: Same as above, removed the others instead.

File: shared/js/background/classes/tab.es6.js [hunk 5]
```diff
@@ -29,37 +29,187 @@ const Site = require('./site.es6')
...
+        this._tabState.setValue('status', value)
+    }
+
+    get statusCode () {
+        return this._tabState.statusCode
+    }
 
-        /** @type {null | import('../events/referrer-trimming').Referrer} */
-        this.referrer = null
+    set statusCode (value) {
```
[the remark this answers is not in the document]
Comment (jonathanKingston) [reply shared/js/background/classes/tab.es6.js:211]: Same as above, removed the others instead.

File: shared/js/background/classes/tab-state.js (hunk 1 above)
Comment (jonathanKingston) [reply shared/js/background/classes/tab-state.js:123]: I think they should be rare enough to be safe to ignore, largely the tab state is safe to ignore a few writes to and potentially risk a bit of data loss.

I think the only likely cause is storage exceeded which we're currently not handling and likely will be a follow up to see if we should care / how to handle it :)

**Review by shakyShane — APPROVED** [review]
🚀

**Review by shakyShane — APPROVED** [review]

**Pull request merged**